### PR TITLE
Listing deployments from contracts instead of kvstore

### PR DESCRIPTION
**Pull request by AhmedHanafy725**
### Description

most of the workloads are vms and we have 2 different clients to create these workloads, but client client was having his way of saving the deployment data to list them. So to be able to list workload from these clients on the dashboard, they have to use the same metadata stored on the contracts' data.

### Changes

This PR lists the deployments (workloads) based on the contracts' data as this metadata is unified between the clients instead of the kvstore.

### Related Issues

- #23 
- #1389 

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

**Comment by MohamedElmdary** [comment]
- *__Virtual Machine__*
  - Full VM
    - [ ] Dashboard
    - [ ] Grid_Client
  - Micro VM
    - [ ] Dashboard
    - [ ] Grid_Client
- *__Orchestrators__*
  - K8s
    - [ ] Dashboard
    - [ ] Grid_Client
  - Caprover
    - [ ] Dashboard
    - [ ] Grid_Client
- *__Applications__*
  - Peertube
    - [ ] Dashboard
    - [ ] Grid_Client
  - Funkwhale
    - [ ] Dashboard
    - [ ] Grid_Client
  - Mattermost
    - [ ] Dashboard
    - [ ] Grid_Client
  - Discourse
    - [ ] Dashboard
    - [ ] Grid_Client
  - Taiga
    - [ ] Dashboard
    - [ ] Grid_Client
  - Owncloud
    - [ ] Dashboard
    - [ ] Grid_Client
  - Nextcloud
    - [ ] Dashboard
    - [ ] Grid_Client
  - Presearch
    - [ ] Dashboard
    - [ ] Grid_Client
  - Subsquid
    - [ ] Dashboard
    - [ ] Grid_Client
  - Casperlabs
    - [ ] Dashboard
    - [ ] Grid_Client
  - Algorand
    - [ ] Dashboard
    - [ ] Grid_Client
  - Node Pilot
    - [ ] Dashboard
    - [ ] Grid_Client
  - Umbrel
    - [ ] Dashboard
    - [ ] Grid_Client
  - Wordpress
    - [ ] Dashboard
    - [ ] Grid_Client

**Comment by MohamedElmdary** [comment]
Failed to list k8s after create an instance ... investing

File: packages/grid_client/src/clients/tf-grid/contracts.ts [hunk 1]
```diff
@@ -141,6 +146,53 @@ class TFContracts extends Contracts {
...
+                  numberOfPublicIPs
+                }
+              }`;
+      const response = await gqlClient.query(body, {
+        nodeContractsCount,
+      });
+
+      return (response["data"] as GqlContracts).nodeContracts;
+    } catch (err) {
+      (err as Error).message = formatErrorMessage(`Error listing contracts by twin id ${options.twinId}.`, err);
```
Comment (0oM4R) [inline packages/grid_client/src/clients/tf-grid/contracts.ts:191]: should we introduce a new error class and use it here ?

**Comment by 0oM4R** [comment]
[Screencast from 12 فبر, 2024 EET 04:38:10 م.webm](https://github.com/threefoldtech/tfgrid-sdk-ts/assets/62248851/05e2cd87-8c36-4952-b37e-80e51019a1b1)


this deployment is vm with qsfs, the config btn is disabled, but on click it opens the details dialog. 
i guess it should show a tooltip or something to explain the reason of disable it

**Comment by 0oM4R** [comment]
i deployed k8s using grid client, tfkvstore as backend storage, but it appears with `List All deployment` off 
image



json 
```json
{
  "masters": [
    {
      "version": 0,
      "contractId": 86372,
      "nodeId": 11,
      "name": "master",
      "created": 1707750417,
      "status": "ok",
      "message": "",
      "flist": "https://hub.grid.tf/tf-official-apps/threefoldtech-k3s-latest.flist",
      "publicIP": null,
      "interfaces": [
        {
          "network": "monNetwork",
          "ip": "10.238.4.2"
        }
      ],
      "capacity": {
        "cpu": 1,
        "memory": 1024
      },
      "mounts": [
        {
          "name": "master_disk",
          "mountPoint": "/mnt/data",
          "size": 1073741824,
          "state": "ok",
          "message": ""
        }
      ],
      "env": {
        "SSH_KEY": "acoustic apology father noble strike brass print denial language effort measure carbon",
        "K3S_TOKEN": "secret",
        "K3S_DATA_DIR": "/mnt/data",
        "K3S_FLANNEL_IFACE": "eth0",
        "K3S_NODE_NAME": "master",
        "K3S_URL": ""
      },
      "entrypoint": "/sbin/zinit init",
      "metadata": "{\"type\":\"kubernetes\",\"name\":\"testk8s\",\"projectName\":\"\"}",
      "description": "test deploying k8s via ts grid3 client",
      "rootfs_size": 0,
      "corex": false,
      "gpu": [],
      "billing": "No Data Available"
    }
  ],
  "workers": [
    {
      "version": 0,
      "contractId": 86372,
      "nodeId": 11,
      "name": "worker",
      "created": 1707750421,
      "status": "ok",
      "message": "",
      "flist": "https://hub.grid.tf/tf-official-apps/threefoldtech-k3s-latest.flist",
      "publicIP": null,
      "interfaces": [
        {
          "network": "monNetwork",
          "ip": "10.238.4.3"
        }
      ],
      "capacity": {
        "cpu": 1,
        "memory": 1024
      },
      "mounts": [
        {
          "name": "worker_disk",
          "mountPoint": "/mnt/data",
          "size": 1073741824,
          "state": "ok",
          "message": ""
        }
      ],
      "env": {
        "SSH_KEY": "acoustic apology father noble strike brass print denial language effort measure carbon",
        "K3S_TOKEN": "secret",
        "K3S_DATA_DIR": "/mnt/data",
        "K3S_FLANNEL_IFACE": "eth0",
        "K3S_NODE_NAME": "worker",
        "K3S_URL": "https://10.238.4.2:6443"
      },
      "entrypoint": "/sbin/zinit init",
      "metadata": "{\"type\":\"kubernetes\",\"name\":\"testk8s\",\"projectName\":\"\"}",
      "description": "test deploying k8s via ts grid3 client",
      "rootfs_size": 0,
      "corex": false,
      "gpu": []
    }
  ],
  "deploymentName": "testk8s",
  "projectName": "testk8s",
  "wireguard": "[Interface]\nAddress = 100.64.238.3/32\nPrivateKey = a6okvni0GdrYNQGIDeIptzfMA+thbuM0P8WZwc3mqLk=\n\n[Peer]\nPublicKey = Vu+VCxj1vTm9aYsnGyKsUpvyMIDS4OrSXwJDuigZVl0=\nAllowedIPs = 10.238.0.0/16, 100.64.238.0/32\nPersistentKeepalive = 25\nEndpoint = 185.206.122.31:29439",
  "name": "testk8s",
  "ipv4": "None",
  "ipv6": "None",
  "planetary": "None",
  "workersLength": 1,
  "billing": "No Data Available",
  "created": 1707750417
}
```

**Comment by AhmedHanafy725** [comment]
> also please confirm, that can i manage the deployments from outside the dashboard or no?

no, you can't only listing for now

**Comment by 0oM4R** [comment]
vm `vml53eg` listed twice on enabling `list all deployments`, and also listed once when it is disabled,
this vm deployed from [staging](https://staging.dashboard.dev.grid.tf/#/deploy/virtual-machines/micro-virtual-machine/) 
image

**Comment by 0oM4R** [comment]
i couldn't list the k8s deployment the comes from terraform
image
image

```josn
{
  "version": 0,
  "twin_id": 201,
  "contract_id": 86458,
  "metadata": "{\"type\":\"kubernetes\",\"name\":\"mr\",\"projectName\":\"Kubernetes\"}",
  "description": "",
  "expiration": 0,
  "signature_requirement": {
    "requests": [
      {
        "twin_id": 201,
        "required": false,
        "weight": 1
      }
    ],
    "weight_required": 1,
    "signatures": [
      {
        "twin_id": 201,
        "signature": "8e58e8e3d0eac07da542d7ad49293c66c0cab88249e62ec6fe6a2ca6f4112a4ab528e2cb6594c401ea1ab8ac7cdeadfb4a5d242d8c653ce0e9374a142093a581",
        "signature_type": "sr25519"
      }
    ],
    "signature_style": ""
  },
  "workloads": [
    {
      "version": 0,
      "name": "w0",
      "type": "zmachine",
      "data": {
        "flist": "https://hub.grid.tf/tf-official-apps/threefoldtech-k3s-latest.flist",
        "network": {
          "public_ip": "",
          "planetary": false,
          "interfaces": [
            {
              "network": "myk8s",
              "ip": "10.1.4.2"
            }
          ]
        },
        "size": 0,
        "compute_capacity": {
          "cpu": 2,
          "memory": 1073741824
        },
        "mounts": [
          {
            "name": "w0disk",
            "mountpoint": "/mydisk"
          }
        ],
        "entrypoint": "/sbin/zinit init",
        "env": {
          "K3S_DATA_DIR": "/mydisk",
          "K3S_FLANNEL_IFACE": "eth0",
          "K3S_NODE_NAME": "w0",
          "K3S_TOKEN": "12345678910122",
          "K3S_URL": "https://10.1.3.2:6443",
          "SSH_KEY": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDOGC+q4UFmijIba/GOP3wbuGGEtd+AwKxw3sFcgcz/AV5dri6GSI36bQEC3Kv+GatRRI3y0PqyciqKmZrnjfON//LKd/5kQOLtz8VHY6FCan3aj+jtpB8r85bIJgw7PyMR7cyJ47RDQsIndehcQLiJcnmOmmmYbLITGoJXFSzlNMN1NlJaCXQszX+tE8kzcse+0xqB4CM9pJNfEnBgAVCygPZeiLgU/mytear+LtGWMoMl9Bkx2CzmPQfWPsq4Svh4Nb3I9BIfnLOB0o3bEM3LqlcuNkS46nPf0ehHIqSOHYpwtL9ryH3PdNpnvxyZD3a7EIbC1fINUsA2W50hoJINm/Tu04MF9GDqeWO1ZCj/lfUPTm+XgaWk0shUlcXJ4iqMprNgD2E3tNNEVMywJ2tJacdl8dJzQG4M+/3TqUx4vpgc1fFJSPSW0SxvPdZ2Gmqh4hGMgcg0kg+Gq4m/u6iW6q7QpdWe/ODLm7zbQ9hodtKVWFMUuugF49ahewyGSG0= khaledyoussef@Khaleds-Air\n"
        },
        "corex": false
      },
      "metadata": "",
      "description": "",
      "result": {
        "created": 1707756143,
        "state": "ok",
        "message": "",
        "data": {
          "id": "201-86458-w0",
          "ip": "10.1.4.2",
          "planetary_ip": "",
          "console_url": "10.1.4.0:20002"
        }
      }
    },
    {
      "version": 0,
      "name": "w0disk",
      "type": "zmount",
      "data": {
        "size": 2147483648
      },
      "metadata": "",
      "description": "",
      "result": {
        "created": 1707756139,
        "state": "ok",
        "message": "",
        "data": {
          "volume_id": "201-86458-w0disk"
        }
      }
    }
  ]
}
```

**Comment by zaelgohary** [comment]
The "Show all deployments" action might seem unclear to users. Adding a tooltip could provide more clarity on its purpose.

image

**Comment by MohamedElmdary** [comment]
> The "Show all deployments" action might seem unclear to users. Adding a tooltip could provide more clarity on its purpose.
> 
> image

can  you recommend a tooltip for clearing it?

**Comment by zaelgohary** [comment]
Is it intended that the "Show All Deployments" button is not shown in Micro VM? AFAIK, we can deploy micro VMs outside the Dashboard.

image

**Comment by MohamedElmdary** [comment]
> Is it intended that the "Show All Deployments" button is not shown in Micro VM? AFAIK, we can deploy micro VMs outside the Dashboard.
> 
> image

it's only shown when there is at least 1 vm from outside playground

**Comment by zaelgohary** [comment]
> > The "Show all deployments" action might seem unclear to users. Adding a tooltip could provide more clarity on its purpose.
> > image
> 
> can you recommend a tooltip for clearing it?

"List all deployments, including those created outside the Dashboard."

**Comment by 0oM4R** [comment]
> > Is it intended that the "Show All Deployments" button is not shown in Micro VM? AFAIK, we can deploy micro VMs outside the Dashboard.
> > image
> 
> it's only shown when there is at least 1 vm from outside playground

@MohamedElmdary 
i don't have any k8s deployments, but the button is shown here:
[Screencast from 13 فبر, 2024 EET 11:45:37 ص.webm](https://github.com/threefoldtech/tfgrid-sdk-ts/assets/62248851/50e33315-c708-4cfd-ada3-55696fc02258)

File: packages/playground/src/utils/load_deployment.ts [hunk 2]
```diff
@@ -151,7 +151,7 @@ export async function loadK8s(grid: GridClient) {
 
   const projectName = grid.clientOptions.projectName;
   const grids = (await Promise.all(
-    clusters.map(n => getGrid(grid.clientOptions, `${projectName}/${n}`)),
+    clusters.map(n => getGrid(grid.clientOptions, projectName ? `${projectName}/${n}` : n)),
```
[the remark this answers is not in the document]
Comment (MohamedElmdary) [reply packages/playground/src/utils/load_deployment.ts:154]: `projectName` can be  `""` so we are omitting the leading `/` if there is no `projectName`

File: packages/playground/src/weblets/tf_deployment_list.vue [hunk 3]
```diff
@@ -306,7 +311,13 @@
...
-            <IconActionBtn icon="mdi-cog" tooltip="Manage Workers" @click="dialog = item.value.deploymentName" />
+
+            <IconActionBtn
+              icon="mdi-cog"
+              :disabled="item.value.fromAnotherClient"
+              tooltip="Manage Workers"
+              @click="dialog = item.value.deploymentName"
+            />
 
             <ManageK8SWorkerDialog
```
[the remark this answers is not in the document]
Comment (MohamedElmdary) [reply packages/playground/src/weblets/tf_deployment_list.vue:322]: only `micro vm` and `k8s` affected with this pr ... any caprover instance(deployment) will be shown in micro vm tab

**Comment by MohamedElmdary** [comment]
> > > Is it intended that the "Show All Deployments" button is not shown in Micro VM? AFAIK, we can deploy micro VMs outside the Dashboard.
> > > image
> > 
> > 
> > it's only shown when there is at least 1 vm from outside playground
> 
> This deployment was created yesterday from the dashboard and the button is shown.
> 
> image

This is because we have issue related to k8s
https://github.com/threefoldtech/tfgrid-sdk-ts/issues/2158

File: packages/playground/src/weblets/tf_deployment_list.vue [hunk 4]
```diff
@@ -42,7 +42,12 @@
             @click="openDialog(tabs[activeTab].value, item)"
           />
 
-          <IconActionBtn icon="mdi-cog" tooltip="Manage Domains" @click="dialog = item.value.deploymentName" />
+          <IconActionBtn
+            icon="mdi-cog"
+            tooltip="Manage Domains"
+            :disabled="item.value.fromAnotherClient"
+            @click="dialog = item.value.deploymentName"
+          />
```
[the remark this answers is not in the document]
Comment (MohamedElmdary) [reply packages/playground/src/weblets/tf_deployment_list.vue:51]: only micro vm and k8s affected with this pr

**Comment by MohamedElmdary** [comment]
> > > Is it intended that the "Show All Deployments" button is not shown in Micro VM? AFAIK, we can deploy micro VMs outside the Dashboard.
> > > image
> > 
> > 
> > it's only shown when there is at least 1 vm from outside playground
> 
> @MohamedElmdary i don't have any k8s deployments, but the button is shown here:
> 
>  Screencast.from.13.2024.EET.11.45.37.webm

This is because we have issue related to k8s
https://github.com/threefoldtech/tfgrid-sdk-ts/issues/2158

**Comment by zaelgohary** [comment]
Switch should be disabled if deployments are loading.


[show_switch.webm](https://github.com/threefoldtech/tfgrid-sdk-ts/assets/40770501/ad9723b5-f309-41aa-8133-f706b0fe2e80)

**Comment by MohamedElmdary** [comment]
> Switch should be disabled if deployments are loading.
> 
>  show_switch.webm

it's ok but this switch doesn't affect/ affected by loading  so should we still disable it?

**Comment by 0oM4R** [comment]
I can't list k8s from GridClient with the same secret

image


```json
{
  "version": 0,
  "twin_id": 800,
  "contract_id": 86845,
  "metadata": "{\"type\":\"kubernetes\",\"name\":\"testk8s\",\"projectName\":\"\"}",
  "description": "test deploying k8s via ts grid3 client",
  "expiration": 0,
  "signature_requirement": {
    "requests": [
      {
        "twin_id": 800,
        "required": false,
        "weight": 1
      }
    ],
    "weight_required": 1,
    "signatures": [
      {
        "twin_id": 800,
        "signature": "de174d7fe927fbb4459e42798c0d039b623bff18d1da66c217794545e192cb753dad3ffb57f1df4badb28af51472a66090c95023cac0955acac92a6738c84584",
        "signature_type": "sr25519"
      }
    ],
    "signature_style": ""
  },
  "workloads": [
    {
      "version": 0,
      "name": "master_disk",
      "type": "zmount",
      "data": {
        "size": 1073741824
      },
      "metadata": "{\"type\":\"kubernetes\",\"name\":\"testk8s\",\"projectName\":\"\"}",
      "description": "test deploying k8s via ts grid3 client",
      "result": {
        "created": 1707819448,
        "state": "ok",
        "message": "",
        "data": {
          "volume_id": "800-86845-master_disk"
        }
      }
    },
    {
      "version": 0,
      "name": "worker",
      "type": "zmachine",
      "data": {
        "flist": "https://hub.grid.tf/tf-official-apps/threefoldtech-k3s-latest.flist",
        "network": {
          "public_ip": "",
          "interfaces": [
            {
              "network": "monNetwork",
              "ip": "10.238.4.3"
            }
          ],
          "planetary": true
        },
        "size": 0,
        "compute_capacity": {
          "cpu": 1,
          "memory": 1073741824
        },
        "mounts": [
          {
            "name": "worker_disk",
            "mountpoint": "/mnt/data"
          }
        ],
        "entrypoint": "/sbin/zinit init",
        "env": {
          "SSH_KEY": "acoustic apology father noble strike brass print denial language effort measure carbon",
          "K3S_TOKEN": "secret",
          "K3S_DATA_DIR": "/mnt/data",
          "K3S_FLANNEL_IFACE": "eth0",
          "K3S_NODE_NAME": "worker",
          "K3S_URL": "https://10.238.4.2:6443"
        },
        "corex": false,
        "gpu": []
      },
      "metadata": "{\"type\":\"kubernetes\",\"name\":\"testk8s\",\"projectName\":\"\"}",
      "description": "test deploying k8s via ts grid3 client",
      "result": {
        "created": 1707819456,
        "state": "ok",
        "message": "",
        "data": {
          "id": "800-86845-worker",
          "ip": "10.238.4.3",
          "planetary_ip": "302:9e63:7d43:b742:553:98d4:1bc:dee0",
          "console_url": "10.238.4.0:20003"
        }
      }
    },
    {
      "version": 0,
      "name": "master",
      "type": "zmachine",
      "data": {
        "flist": "https://hub.grid.tf/tf-official-apps/threefoldtech-k3s-latest.flist",
        "network": {
          "public_ip": "",
          "interfaces": [
            {
              "network": "monNetwork",
              "ip": "10.238.4.2"
            }
          ],
          "planetary": true
        },
        "size": 0,
        "compute_capacity": {
          "cpu": 1,
          "memory": 1073741824
        },
        "mounts": [
          {
            "name": "master_disk",
            "mountpoint": "/mnt/data"
          }
        ],
        "entrypoint": "/sbin/zinit init",
        "env": {
          "SSH_KEY": "acoustic apology father noble strike brass print denial language effort measure carbon",
          "K3S_TOKEN": "secret",
          "K3S_DATA_DIR": "/mnt/data",
          "K3S_FLANNEL_IFACE": "eth0",
          "K3S_NODE_NAME": "master",
          "K3S_URL": ""
        },
        "corex": false,
        "gpu": []
      },
      "metadata": "{\"type\":\"kubernetes\",\"name\":\"testk8s\",\"projectName\":\"\"}",
      "description": "test deploying k8s via ts grid3 client",
      "result": {
        "created": 1707819452,
        "state": "ok",
        "message": "",
        "data": {
          "id": "800-86845-master",
          "ip": "10.238.4.2",
          "planetary_ip": "302:9e63:7d43:b742:7e54:993a:46b0:171a",
          "console_url": "10.238.4.0:20002"
        }
      }
    },
    {
      "version": 0,
      "name": "worker_disk",
      "type": "zmount",
      "data": {
        "size": 1073741824
      },
      "metadata": "{\"type\":\"kubernetes\",\"name\":\"testk8s\",\"projectName\":\"\"}",
      "description": "test deploying k8s via ts grid3 client",
      "result": {
        "created": 1707819448,
        "state": "ok",
        "message": "",
        "data": {
          "volume_id": "800-86845-worker_disk"
        }
      }
    }
  ]
}

```

**Comment by MohamedElmdary** [comment]
> Member

checking it

**Comment by zaelgohary** [comment]
I can't list deployments created from Terraform. This account has VM, K8s, Presearch and Mattermost deployments.

image

**Comment by MohamedElmdary** [comment]
> I can't list deployments created from Terraform. This account has VM, K8s, Presearch and Mattermost deployments.
> 
> image
> 
> image
> 
> image

Should be fixed now in k8s and machines
image

**Comment by zaelgohary** [comment]
> > I can't list deployments created from Terraform. This account has VM, K8s, Presearch and Mattermost deployments.
> > image
> > image
> > image
> 
> Should be fixed now in k8s and machines image

Mattermost is still missing.

**Comment by MohamedElmdary** [comment]
> > > I can't list deployments created from Terraform. This account has VM, K8s, Presearch and Mattermost deployments.
> > > image
> > > image
> > > image
> > 
> > 
> > Should be fixed now in k8s and machines image
> 
> Mattermost is still missing.

My bad forgot to mention that mattermost instance called `ashrafmattermost` on node 3 which is down right now

**Comment by MohamedElmdary** [comment]
> List all deployments, including those created outside the Dashboard.

Added
image

**Comment by 0oM4R** [comment]
created vm and k8s using gridclient with secretStore:
and deleted them, no errors but the network contracts not deleted 
image
image
image

**Comment by 0oM4R** [comment]
after deleting k8s instance that deployed from grid client it shows the loading deployment failed alert
image

**Comment by 0oM4R** [comment]
[Uploading Screencast from 13 فبر, 2024 EET 05:44:51 م.webm…]()

**Pull request merged**